### PR TITLE
feat: support loader-runner and pitching

### LINKED ISSUE #1750: Loader Identifier support
Loader config allows user to pass an identifier as the key of each loader, the absolute path is preferred though.

### LINKED ISSUE #1751: Support resolving inline loader from the native side
Inline loaders are different, as these should be resolved as functions from the native side.

Related: https://github.com/web-infra-dev/rspack/issues/3737

### LINKED ISSUE #2394: [Feature Request]: pitching loader support
https://webpack.js.org/api/loaders/#pitching-loader

### PR DESCRIPTION
**Pull request by h-a-n-a**
## Related issue (if exists)

closes #2394 
closes #1750
closes #1751

## Summary

This PR refactors the loader architecture and adapts webpack's loader-runner to rspack.
Pitching loader is supported for both Rust and JS side from now on.

The old architecture is a quite simple version, which only supports loaders for normal stage.
Pitching loader does not put into consideration. The basic concept of the old version is to
convert the normal loader to a _**native function**_ which can be called from the Rust side.
Furthermore, for performance reason, Rspack also composes loaders from the JS side to 
mitigate the performance issue of Node/Rust communications.

In this new architecture, loaders will not be converted directly into _**native functions**_.
Instead, it is almost the same with how webpack's loader-runner resolves its loaders, by
leveraging the identifier. Every time Rspack wants to invoke a JS loader, the identifiers will
be passed to the handler passed by Node side to process. The implementation also **keeps 
the feature of composing JS loaders** for performance reason.

See https://github.com/web-infra-dev/rspack/discussions/2846 for more details on the architecture.

Tracking PRs:
- [X] https://github.com/web-infra-dev/rspack/pull/2789

### <samp>🤖 Generated by Copilot at 2c6ea7c</samp>

This pull request refactors the `rspack_core` and `rspack_loader_runner` crates to use a new `ResolverFactory` abstraction for resolving module dependencies and a new `run_loaders` function for running loaders. It also adds support for more complex resource URLs, loaders, and modules, and implements hot module replacement (HMR) for modules. It removes unused and outdated code and simplifies the code with the `derivative` crate.

<details open=true>
  <summary><h2>Walkthrough</h2></summary>

### <samp>🤖 Generated by Copilot at 2c6ea7c</samp>

*  Remove the `LoaderRunnerRunner` type and replace it with the `ResolverFactory` type and the `run_loaders` function to simplify the loader running and resolving logic ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L43-R45), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L80-R80), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L115-R115), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L135-R135), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5eb759d814281903aa0355174051f134df9444a46e9a1ce332cad27eed896215L630-R630), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-269c100a13b91351b9f928955094fdd7641e93f1f3062c0de339fd041fb4e763L153-R153), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L18-R18), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L32-R29), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L54-L58), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L68-R65), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-3fbb9f7dfbadceab7b0c89038c54de9851f25ae957a1f91aea05b0e46da2b209L103-R95), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L6-R10), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L223-R223), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L242-R242), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L257-R261), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-dbe245f1bef02146810f8c3ec8aad0c45cfdd8acc7ed605743ae9f8ec30bb935L3-R5), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-dbe245f1bef02146810f8c3ec8aad0c45cfdd8acc7ed605743ae9f8ec30bb935L19-R15), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-bb344009ce65860c820c92132f4f7d7653f6d19c7260699cd3d27b58818ddd90L505-R518))
* Add the `CompilerContext` type and the `BuildContext` type to hold the common compiler options and the `ResolverFactory` instance that are used by different modules and loaders ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-5354812b8e30bff2ce2e8f7dfc292bdee3bad5022b1efc6fc1b85ab5a084eb14L6-R10), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-6d6ee10fb57dcb86a4a0d44d272155b4466773c75b8a006f4c2b4bbe6209efa0L12-R19))
* Add the `LoaderRunnerPluginProcessResource` type and use it in the `NormalModule` struct to support plugin hooks for the loader runner ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-8390e2e37223b1f6ac009194ef6d1a9272f9f15e2430da928b938d7d19723822L7-R7), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-bb344009ce65860c820c92132f4f7d7653f6d19c7260699cd3d27b58818ddd90L32-R35))
* Add the `resource_fragment`, `resource_query`, and `resource_path` fields to the `ResourceData` struct to support more complex resource URLs and parameters ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-4e77827441656f6c0f2713ed92f4f4a58029006543ba5e686c842d1b28de6c0eL97-R99))
* Modify the `identifier` method of the `NormalModuleFactory` struct to use the `identifier` method of the `Loader` trait instead of the `name` method to support more complex and dynamic loaders and modules ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-4e77827441656f6c0f2713ed92f4f4a58029006543ba5e686c842d1b28de6c0eL188-R190))
* Add the `derivative` crate and the `Derivative` derive macro to customize the `Debug` implementation for some structs and enums in the `rspack_core` crate ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-b15514bda6ea6c4fd02c5fe3c17a76a128bae9fc869d5245b0d2a109635e91f8R15), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-bb344009ce65860c820c92132f4f7d7653f6d19c7260699cd3d27b58818ddd90L14-R19), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-bb344009ce65860c820c92132f4f7d7653f6d19c7260699cd3d27b58818ddd90L318-R320), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-bb344009ce65860c820c92132f4f7d7653f6d19c7260699cd3d27b58818ddd90R336), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-06c579a7786ea5f17871a0f626ca314d0a41ffb8ba25152ce1d7661e6d1f3d8eR7), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-06c579a7786ea5f17871a0f626ca314d0a41ffb8ba25152ce1d7661e6d1f3d8eL122-R124), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-06c579a7786ea5f17871a0f626ca314d0a41ffb8ba25152ce1d7661e6d1f3d8eR140))
* Rename the `rspack_loader_runner2` crate to `rspack_loader_runner` and update the `rspack_core` crate to use the new version ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-4168cbe766389af33fc1d0840ddf230a49d5e3d77b4cfa9bcfcae55d293238dcL4-R4), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-b15514bda6ea6c4fd02c5fe3c17a76a128bae9fc869d5245b0d2a109635e91f8L33-R34))
* Delete the old version of the `rspack_loader_runner` crate and its files ([link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-531b65ecb917d7a39a15bb06a601f70657c07a29a5b8e0e007e263e9d0dd1f06), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-a333011d485331a49bdc52c0b20fa4dda32b044f16efce5c58c949694ea459db), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-10bb652578189ccda5abb765d4a7451e448519bbe93aa32c09ea2a5fa292ba3c), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-4854d56fe47c50e9bc71d9007f1bc701af81892a7b7d5637a9d86215bf53b37f), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-acead8fd8a1389c313f342ed0162ed7712cc3a217c4374f0baaaa9c237fccedd), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-16cb150fcaca739686ff5867e4ee40a8fc401ff949fe0f90cf74a7d768b39381), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-cacb920156b22fc5231261ef12e609129def0bc11da357e321e9559f65ed711f), [link](https://github.com/web-infra-dev/rspack/pull/2808/files?diff=unified&w=0#diff-6614b8a937f22f3e4b4a60436155b93b7cab3260a84bc8ca17e1c26d351cfcdf))

</details>

### CODE REVIEW

**Comment by hardfist** [comment]
it would be better if we have a example about pitching loader

File: crates/rspack_binding_options/src/options/raw_module/js_loader.rs [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+  rspack_core::{Loader, LoaderContext, LoaderRunnerContext},
+  rspack_error::internal_error,
+  rspack_identifier::{Identifiable, Identifier},
+  rspack_napi_shared::threadsafe_function::{ThreadsafeFunction, ThreadsafeFunctionCallMode},
+  rspack_napi_shared::{NapiResultExt, NAPI_ENV},
+};
+
+#[napi(object)]
+pub struct JsLoader {
+  /// composed loader name, xx-loader$yy-loader$zz-loader
```
Comment (h-a-n-a) [inline crates/rspack_binding_options/src/options/raw_module/js_loader.rs:16]: Like the previous PR said, the composed JS loaders will now be joined with `$` for better interoperability between the Rust and Node side.

**Comment by hardfist** [comment]
!bench

**Comment by h-a-n-a** [comment]
> it would be better if we have a example about pitching loader

@hardfist Pitching loader is often used with inline loaders to initiate a new loader pipeline like how style-loader did, rspack does not support inline loaders now, I would like to add an example after style-loader or something else is supported.

**Review by ahabhgk — APPROVED** [review]

File: crates/rspack_binding_options/src/options/raw_module/mod.rs [hunk 2]
```diff
@@ -328,258 +323,58 @@ pub struct RawModuleOptions {
...
-                return JsLoaderAdapter::try_from(raw_js_loader).map(|i| Arc::new(i) as BoxLoader);
+    let uses = self
+        .r#use
+        .map(|uses| {
+          uses
+            .into_iter()
+            .map(|rule_use| {
+              #[cfg(feature = "node-api")]
+              {
+                if let Some(raw_js_loader) = rule_use.js_loader {
```
[the remark this answers is not in the document]
Comment (h-a-n-a) [reply crates/rspack_binding_options/src/options/raw_module/mod.rs:343]: Loader creation and loader composition are two different steps.

**Pull request merged**